### PR TITLE
[13.0][IMP+FIX] hr_expense_invoice: Add button to create invoice from expenses + Allow create invoice from invoice_id field

### PR DESCRIPTION
**Pull request by victoralmau**
[IMP] Add button to create invoice from expenses.
[FIX] Allow create invoice from invoice_id field

In these PR https://github.com/OCA/hr-expense/pull/32 we remove it option to create invoice from some expenses but it's a interesting functionallity and with these change add button in expense to allow create invoice with only 1 line.

Please @pedrobaeza and @joao-p-marques can you review it?

@Tecnativa TT29072

### CODE REVIEW

**Comment by victoralmau** [comment]
Any idea to create 1 invoice according to multi-expense with "one click"?
Maybe a use case about it is a sheet example with 10 expenses and we have to create only 1 invoice with 10 lines (line per expense).

File: hr_expense_invoice/models/hr_expense.py [hunk 1]
```diff
@@ -21,6 +23,37 @@ class HrExpense(models.Model):
...
+                    "tax_ids": [(6, 0, self.tax_ids.ids)],
+                },
+            )
+        ]
+        invoice = (
+            self.env["account.move"]
+            .with_context(default_type="in_invoice")
+            .create(
+                {
+                    "type": "in_invoice",
```
Comment (joao-p-marques) [inline hr_expense_invoice/models/hr_expense.py]: Why do you need to specify the type twice? I'd say the correct one would be through the context.

File: hr_expense_invoice/tests/test_hr_expense_invoice.py [hunk 2]
```diff
@@ -232,6 +232,39 @@ def test_2_hr_test_multi_invoices(self):
...
+        self.expense2.action_expense_create_invoice()
+        self.assertTrue(self.expense2.invoice_id)
+        self.assertEqual(self.sheet.invoice_count, 2)
+        # Validate invoices
+        self.expense.invoice_id.partner_id = self.partner
+        self.expense.invoice_id.action_post()
+        self.expense2.invoice_id.partner_id = self.partner
+        self.expense2.invoice_id.action_post()
+        self.sheet.with_context(
+            {"default_expense_line_ids": self.expense.id}
```
Comment (joao-p-marques) [inline hr_expense_invoice/tests/test_hr_expense_invoice.py]: Are you sure this is correct? Isn't `default_expense_line_ids` expecting a list of ids?

**Review by joao-p-marques — COMMENTED** [review]
Code review

**Review by joao-p-marques — APPROVED** [review]

File: hr_expense_invoice/views/hr_expense_views.xml [hunk 3]
```diff
@@ -41,6 +44,18 @@
...
+            <xpath
+                expr="//field[@name='expense_line_ids']/tree/field[@name='total_amount']"
+                position="after"
+            >
+                <field name="sheet_id_state" invisible="1" />
+                <button
+                    name="action_expense_create_invoice"
+                    string="Create Vendor Bill"
+                    type="object"
+                    attrs="{'invisible': ['|', ('invoice_id', '!=', False),('sheet_id_state', '!=', 'approve')]}"
```
Comment (pedrobaeza) [inline hr_expense_invoice/views/hr_expense_views.xml]: Shouldn't you allow it on draft as well?

File: hr_expense_invoice/models/account_move.py [hunk 4]
```diff
@@ -1,11 +1,26 @@
...
     _inherit = "account.move"
 
+    expense_ids = fields.One2many(
+        comodel_name="hr.expense", inverse_name="invoice_id", string="Expenses"
+    )
+
+    @api.constrains("amount_total")
+    def _check_expense_ids(self):
+        for move in self:
+            if move.expense_ids:
```
Comment (pedrobaeza) [inline hr_expense_invoice/models/account_move.py]: Save one indentation level:
```suggestion
        for move in self.filtered("expense_ids"):
```

File: hr_expense_invoice/models/account_move.py [hunk 5]
```diff
@@ -1,11 +1,26 @@
...
+                if sum(move.expense_ids.mapped("total_amount")) != move.amount_total:
+                    raise UserError(
```
Comment (pedrobaeza) [inline hr_expense_invoice/models/account_move.py]: On constraints, it should be `ValidationError`:
```suggestion
                    raise ValidationError(
```

File: hr_expense_invoice/models/account_move.py [hunk 6]
```diff
@@ -1,11 +1,26 @@
...
+                        _("Total amount can't change because some expense is related")
```
Comment (pedrobaeza) [inline hr_expense_invoice/models/account_move.py]:
```suggestion
                        _("You can't change the total amount, as there's an expense linked to this invoice.")
```

File: hr_expense_invoice/models/account_move.py [hunk 7]
```diff
@@ -1,11 +1,26 @@
...
 
+    expense_ids = fields.One2many(
+        comodel_name="hr.expense", inverse_name="invoice_id", string="Expenses"
+    )
+
+    @api.constrains("amount_total")
+    def _check_expense_ids(self):
+        for move in self:
+            if move.expense_ids:
+                if sum(move.expense_ids.mapped("total_amount")) != move.amount_total:
```
Comment (pedrobaeza) [inline hr_expense_invoice/models/account_move.py]: Use `float_compare` for avoiding rounding problems.

File: hr_expense_invoice/models/hr_expense.py [hunk 8]
```diff
@@ -56,5 +89,5 @@ def _onchange_invoice_id(self):
         """
         if self.invoice_id:
             self.quantity = 1
-            self.unit_amount = self.invoice_id.amount_total
             self.tax_ids = [(5,)]
+            self.unit_amount = self.invoice_id.amount_total
```
Comment (pedrobaeza) [inline hr_expense_invoice/models/hr_expense.py:102]: Why this change or order?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: hr_expense_invoice/models/hr_expense.py (hunk 8 above)
Comment (victoralmau) [reply hr_expense_invoice/models/hr_expense.py:102]: It's very important because if assign `unit_amount` before se `tax_ids` to empty, invoice constraint show (`_check_expense_ids`). 
These little change it's necessary to "_Create Vendor Bill_" button in expense work fine.

Comment (pedrobaeza) [reply hr_expense_invoice/models/hr_expense.py:102]: I'm not sure to get you. Are you saying that a constraint raises? Which one? In any case, if so, please put a comment like:
```python
# Assign this amount after removing taxes for avoiding to raise the constraint X
```
and remember always that when something is not obvious, you should put a comment for that.

File: hr_expense_invoice/models/hr_expense.py [hunk 9]
```diff
@@ -10,17 +11,49 @@
...
+            .with_context(default_type="in_invoice")
+            .create(
+                {
+                    "ref": self.reference,
+                    "invoice_date": self.date,
+                    "invoice_line_ids": invoice_lines,
+                }
+            )
+        )
+        self.write({"invoice_id": invoice.id, "tax_ids": False})
```
Comment (pedrobaeza) [inline hr_expense_invoice/models/hr_expense.py]: I miss here quantity = 1 and unit_amount = total invoice

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: hr_expense_invoice/models/hr_expense.py [hunk 10]
```diff
@@ -10,17 +11,56 @@
...
+    def action_expense_create_invoice(self):
+        invoice_lines = [
+            (
+                0,
+                0,
+                {
+                    "product_id": self.product_id.id,
+                    "name": self.name,
+                    "price_unit": self.unit_amount,
+                    "quantity": 1,
```
Comment (pedrobaeza) [inline hr_expense_invoice/models/hr_expense.py]: This change is not correct AFAIK

File: hr_expense_invoice/views/hr_expense_views.xml [hunk 11]
```diff
@@ -5,18 +5,24 @@
...
-                            'default_reference': reference,
-                            'default_date_invoice': date,
-                            'default_invoice_line_ids': [{'product_id': product_id,
+                            'default_ref': reference,
+                            'default_invoice_date': date,
+                            'default_line_ids': [{'product_id': product_id,
                                                           'name': name,
                                                           'price_unit': unit_amount,
-                                                          'quantity': quantity,
+                                                          'quantity': 1,
```
Comment (pedrobaeza) [inline hr_expense_invoice/views/hr_expense_views.xml]: Neither this one

File: hr_expense_invoice/models/hr_expense.py (hunk 10 above)
Comment (victoralmau) [reply hr_expense_invoice/models/hr_expense.py]: These changes it's about data consistency .
If we change `quantity=1` in `invoice_id` onchange, IMO it's correct to set exactly the same in other use cases.
What do you think about it?

Comment (pedrobaeza) [reply hr_expense_invoice/models/hr_expense.py]: No, it isn't. Starting with, you are not multiplying the unit_amount * quantity in the price_unit, but the question is the consistency is needed from invoice to expense, not the reverse. Please leave it as it was.

Comment (victoralmau) [reply hr_expense_invoice/models/hr_expense.py]: Sorry but if leave it as it was these comment https://github.com/OCA/hr-expense/pull/34#discussion_r615237666 will be pending right?

Comment (pedrobaeza) [reply hr_expense_invoice/models/hr_expense.py]: This one still remains incorrect

**Review by pedrobaeza — APPROVED** [review]
